nengo/synapses.py

```diff
@@ -221,123 +197,123 @@ def evaluate(self, frequencies):
-        if y0 is not None:
-            output[:] = y0
-
-        return LinearFilter.NoDen(np.array([1.]), np.array([]), output)
+        A, B, C, D = self._get_ss(dt, method=method)
+        X = np.zeros((A.shape[0],) + shape_out, dtype=dtype)
+
+        if y0 is not None and len(X) > 0:
+            y0 = np.array(y0, copy=False, ndmin=1)
+            X[:] = np.linalg.solve(np.eye(len(A)) - A, B * y0)
+
+        return dict(X=X)
+
+    def make_step(self, shape_in, shape_out, dt, rng, state=None,
+                  method='zoh'):
+        """Returns a `.Step` instance that implements the linear filter."""
+        assert shape_in == shape_out
+        assert state is not None
+
+        A, B, C, D = self._get_ss(dt, method=method)
+        X = state['X']
+        if LinearFilter.NoX.check(A, B, C, D, X):
+            return LinearFilter.NoX(A, B, C, D, X)
+        elif LinearFilter.OneX.check(A, B, C, D, X):
+            return LinearFilter.OneX(A, B, C, D, X)
+        elif LinearFilter.NoD.check(A, B, C, D, X):
+            return LinearFilter.NoD(A, B, C, D, X)
+        else:
+            assert LinearFilter.General.check(A, B, C, D, X)
+            return LinearFilter.General(A, B, C, D, X)
 
     class Step:
         """Abstract base class for LTI filtering step functions."""
-        def __init__(self, num, den, output):
-            self.num = num
-            self.den = den
-            self.output = output
+        def __init__(self, A, B, C, D, X):
+            if not self.check(A, B, C, D, X):
+                raise ValidationError(
+                    "Matrices do not meet the requirements for this Step",
+                    attr='matrices', obj=self)
+            self.A = A
+            self.B = B
+            self.C = C
+            self.D = D
+            self.X = X
 
         def __call__(self, t, signal):
-            raise NotImplementedError("Step functions must implement __call__")
-
-    class NoDen(Step):
-        """An LTI step function for transfer functions with no denominator.
-
-        This step function should be much faster than the equivalent general
-        step function.
-        """
-        def __init__(self, num, den, output):
-            if len(den) > 0:
-                raise ValidationError("'den' must be empty (got length %d)"
-                                      % len(den), attr='den', obj=self)
-            super().__init__(num, den, output)
-            self.b = num[0]
+            raise NotImplementedError("Step object must implement __call__")
+
+        @classmethod
+        def check(cls, A, B, C, D, X):
+            if A.shape[0] == 0:
+                return X.shape[0] == B.shape[0] == 0 and D.shape[0] == 1
+            else:
+                return (A.shape[0] == A.shape[1] == B.shape[0] == C.shape[1]
+                        == X.shape[0]
+                        and A.shape[0] == X.shape[0]
+                        and D.shape[0] == B.shape[1] == 1)
```

Comment:
Might as well also check that `C.shape[0] == D.shape[1] == 1`?

Reply:
No. `D` is not always 2D.

Reply:
Can you give an example? I believe there was a time in an earlier version of scipy (<0.18.0) when it wasn't, but this was fixed after I reported this issue: https://github.com/scipy/scipy/issues/5760.

Reply:
This was on my laptop, and I can't remember what Scipy version I have on there. Could have been on an older version.